### PR TITLE
[3198] Dummy Exchange for goerly network

### PR DESCRIPTION
**Pull request by adamskrodzki**
In order to run tests first
`npx hardhat node`

then (in other terminal)

`npx hardhat test test\goerli-exchange-tests.js --network local`

[Story](https://app.shortcut.com/oazo-apps/story/3198/create-another-type-of-dummyexchange)

### CODE REVIEW

File: contracts/multiply/GoerliDummyExchange.sol [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+import "../utils/SafeMath.sol";
+import "../utils/SafeERC20.sol";
+import "hardhat/console.sol";
+
+contract GoerliDummyExchange {
+  using SafeMath for uint256;
+  using SafeERC20 for IERC20;
+
+  address DAI_ADDRESS = 0x6B175474E89094C44Da98b954EedeAC495271d0F;
+  mapping(address => bool) public WHITELISTED_CALLERS;
```
Comment (johnnieskywalker) [inline contracts/multiply/GoerliDummyExchange.sol:14]: Where are WHITELISTED_CALLERS stored?

File: contracts/multiply/GoerliDummyExchange.sol [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+
+  function _collectFee(address asset, uint256 fromAmount) public returns (uint256) {
+    uint256 feeToTransfer = fromAmount.mul(fee).div(feeBase);
+    IERC20(asset).transferFrom(address(this), feeBeneficiaryAddress, feeToTransfer);
+    emit FeePaid(feeBeneficiaryAddress, feeToTransfer);
+    return fromAmount.sub(feeToTransfer);
+  }
+
+  // uses the same interface as default Exchange contract
+  function swapDaiForToken(
```
Comment (johnnieskywalker) [inline contracts/multiply/GoerliDummyExchange.sol:82]: So interfaces are reused but there's no syntax to explicitly show that this is the same type of an interface in default and dummy exchange ?

File: test/goerli-exchange-tests.js [hunk 3]
```diff
@@ -0,0 +1,393 @@
...
+        lessThanTheTransferAmount = convertToBigNumber(amountInWei).minus(5);
+
+        await WETH.approve(exchange.address, lessThanTheTransferAmount.toFixed(0))
+      })
+
+      this.afterEach(async function () {
+        await provider.send('evm_revert', [localSnapshotId])
+      })
+
+      it('should throw an error and not exchange anything', async function () {
```
Comment (johnnieskywalker) [inline test/goerli-exchange-tests.js:220]: This error will be thrown because o value 0 right ?

File: contracts/multiply/GoerliDummyExchange.sol [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+    require(WHITELISTED_CALLERS[msg.sender], "Exchange / Unauthorized Caller.");
+    _;
+  }
+
+  constructor(address _beneficiary, uint8 _fee, uint8 _slippage, address _dai, address authorisedCaller){
+    feeBeneficiaryAddress = _beneficiary;
+    fee = _fee;
+    slippage = _slippage;
+    DAI_ADDRESS = _dai;
+    WHITELISTED_CALLERS[authorisedCaller] = true;
```
Comment (johnnieskywalker) [inline contracts/multiply/GoerliDummyExchange.sol:40]: We have map of `WHITELISTED_CALLERS` however in constructor we pass only one `authorisedCaller`

File: contracts/multiply/GoerliDummyExchange.sol [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+import "../interfaces/IERC20.sol";
+import "../utils/SafeMath.sol";
+import "../utils/SafeERC20.sol";
+import "hardhat/console.sol";
+
+contract GoerliDummyExchange {
+  using SafeMath for uint256;
+  using SafeERC20 for IERC20;
+
+  address DAI_ADDRESS = 0x6B175474E89094C44Da98b954EedeAC495271d0F;
```
Comment (fenris85) [inline contracts/multiply/GoerliDummyExchange.sol:13]: You could hardcode goerli Dai address and do not pass it in the constructor.

File: contracts/multiply/GoerliDummyExchange.sol [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+    uint256 amountIn,
+    uint256 amountOut
+  );
+  
+  modifier onlyAuthorized() {
+    require(WHITELISTED_CALLERS[msg.sender], "Exchange / Unauthorized Caller.");
+    _;
+  }
+
+  constructor(address _beneficiary, uint8 _fee, uint8 _slippage, address _dai, address authorisedCaller){
```
Comment (fenris85) [inline contracts/multiply/GoerliDummyExchange.sol:35]: Could we get rid of the slippage in this contract at all?

File: test/goerli-exchange-tests.js (hunk 3 above)
Comment (adamskrodzki) [reply test/goerli-exchange-tests.js:220]: actually because of approval of lessThanTheTransferAmount in beforeEach

**Pull request merged**